### PR TITLE
WasmFS JS API: Implement mkdev

### PR DESCRIPTION
**Pull request by jameshu15869**
This PR implements `FS.mkdev` and it's helpers (`registerDevice`, `makedev`, `major`, and `minor`). Existing tests were modified to pass with WasmFS. 

Currently we assign empty methods to any of the required JS_IMPL backend methods that the user does not pass in to prevent errors. Would it be better to instead let the error happen, forcing users to pass in all 5 methods regardless of if they are empty?

In `syscalls.cpp` it looks like there was a minor bug with polling that wasn't caught. Using either `RDONLY` or `RDWR` does not affect the test in legacy mode, but `RDONLY` will cause the test to fail in WasmFS mode. Normally I'd assume this should go in a separate PR, but the fix doesn't seem to make sense on its own is important to helping the tests pass with WasmFS. Should I still move it to a separate PR?

In `test_64bit.c` I'm unsure if we can get a file's position in JS using the function signature requested by `_wasmfs_jsimpl_read`, which doesn't include the position that the old API had. Is there a way to do this?

### CODE REVIEW

File: src/library_wasmfs.js [hunk 1]
```diff
@@ -392,7 +395,60 @@ FS.createPreloadedFile = FS_createPreloadedFile;
...
+        ops.write = defaultOps.write;
+      }
+
+      if (typeof ops.read === 'undefined') {
+        ops.read = defaultOps.read;
+      }
+
+      if (typeof ops.getSize === 'undefined') {
+        ops.getSize = defaultOps.getSize;
+      }
```
Comment (kripken) [inline src/library_wasmfs.js:430]: A more compact way to do this is to use [`Object.assign`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/assign).

File: src/library_wasmfs.js [hunk 2]
```diff
@@ -392,7 +395,60 @@ FS.createPreloadedFile = FS_createPreloadedFile;
         return __wasmfs_mknod(pathBuffer, mode, dev);
       }));
     },
-    // TODO: mkdev
+    major: (dev) => ((dev) >> 8),
+    minor: (dev) => ((dev) & 0xff),
```
Comment (kripken) [inline src/library_wasmfs.js:399]: Are major and minor used?

File: system/lib/wasmfs/js_api.cpp [hunk 3]
```diff
@@ -270,6 +270,10 @@ int _wasmfs_lstat(char* path, struct stat* statBuf) {
   return __syscall_lstat64((intptr_t)path, (intptr_t)statBuf);
 }
 
+int _wasmfs_mkdev(char* path, mode_t mode, backend_t backend) {
+  return wasmfs_create_file(path, mode, backend);
```
Comment (kripken) [inline system/lib/wasmfs/js_api.cpp:274]: Ah, nice this is so simple :smile: 

Actually, perhaps we can just use the existing `wasmfs_create_file`, and not add this?

File: test/core/test_poll.c [hunk 4]
```diff
@@ -9,14 +9,23 @@
...
+    FS.registerDevice(dummy_device, {});
+
+    FS.createDataFile('/', 'file', 'abcdef', true, true, false);
+    FS.mkdev('/device', dummy_device);
+  );
+  
   struct pollfd multi[5];
-  multi[0].fd = open("/file", O_RDONLY, 0777);
-  multi[1].fd = open("/device", O_RDONLY, 0777);
+  multi[0].fd = open("/file", O_RDWR, 0777);
```
Comment (kripken) [inline test/core/test_poll.c:24]: Is this due to the poll change?

**Review by kripken — COMMENTED** [review]
Nice work!

> Currently we assign empty methods to any of the required JS_IMPL backend methods that the user does not pass in to prevent errors. Would it be better to instead let the error happen, forcing users to pass in all 5 methods regardless of if they are empty?

We should error, I think, as there isn't an obvious default behavior to do for, say, a write, if the user didn't say how to actually write. But maybe we should make the error happen at runtime? That is, if the user doesn't provide a write method, that's fine, but if we actually need to write then the program exits with an error on that.

> In syscalls.cpp it looks like there was a minor bug with polling that wasn't caught. Using either RDONLY or RDWR does not affect the test in legacy mode, but RDONLY will cause the test to fail in WasmFS mode. Normally I'd assume this should go in a separate PR, but the fix doesn't seem to make sense on its own is important to helping the tests pass with WasmFS. Should I still move it to a separate PR?

Makes sense to me to leave it here, yeah, as the other PR couldn't get the test to pass yet anyhow.

> In test_64bit.c I'm unsure if we can get a file's position in JS using the function signature requested by _wasmfs_jsimpl_read, which doesn't include the position that the old API had. Is there a way to do this?

Not sure I follow. The read method does say at what place to read, I think - isn't that the only position we need?

File: src/library_wasmfs.js (hunk 2 above)
Comment (jameshu15869) [reply src/library_wasmfs.js:399]: Not in any of the tests, but they were in the old API and were mentioned in the documentation, so I thought it might be important for compatibility. Other than that, there doesn't seem to be any usage that I could think of.

File: test/core/test_poll.c (hunk 4 above)
Comment (jameshu15869) [reply test/core/test_poll.c:24]: Yeah, it looks like prior to the poll change WasmFS would think none of the flags were set because it was checking if `flags == O_RDONLY` instead of checking that `flags & O_ACCESSMODE == O_RDONLY`, etc.

**Comment by jameshu15869** [comment]
> Not sure I follow. The read method does say at what place to read, I think - isn't that the only position we need?

My mistake, I mixed up `position` and `offset`. The test now passes without hardcoding the position in.

File: src/library_wasmfs.js (hunk 2 above)
Comment (kripken) [reply src/library_wasmfs.js:399]: Ah, in that case let's remove them. They might be dead code that was never removed. If we find a need later we'll add them then.

File: src/library_wasmfs.js [hunk 5]
```diff
@@ -59,9 +59,12 @@ FS.createPreloadedFile = FS_createPreloadedFile;
 #endif
     'malloc',
     'free',
+    'wasmfs_create_jsimpl_backend',
+    '$wasmFS$backends'
 #endif
   ],
   $FS : {
+    devices: {},
```
Comment (kripken) [inline src/library_wasmfs.js:67]: It would be better to keep this off on the side, so it isn't included unnecessarily (as it is, all uses of `FS` will get it, even if they didn't enable `FORCE_FILESYSTEM` for example). See e.g. `wasmFSPreloadedFiles` in this file for an example of such a property.

File: test/core/test_poll.c [hunk 6]
```diff
@@ -9,14 +9,35 @@
...
 #include <fcntl.h>
 #include <poll.h>
+#include <emscripten.h>
 
 int main() {
+  EM_ASM(
+    var dummy_device = FS.makedev(64, 0);
+#if WASMFS
+    // WasmFS has different requirements for a device's functions compared
+    // to the legacy API.
```
Comment (kripken) [inline test/core/test_poll.c:19]: Can we reduce these differences? Ideally they would not exist.

Comment (jameshu15869) [reply test/core/test_poll.c:19]: I was thinking about that, but the old FS API allows you to pass custom methods like `llseek` or `open` to devices, whereas those methods aren't accepted/used by the JS IMPL backend (e.g. in `test_64bit.c`). Another difference is that the old FS API allows you to pass an empty set of functions to a device in order to use default behavior (Line 28 of `test_poll.c`), but I think JS IMPL doesn't really have the concept of "default behavior", so we have to at least pass in empty methods for `allocFile`, `freeFile`, etc. 

To keep full compatibility I considered "overriding" the JS methods we provide in `library_wasmfs.js` when a user passes in definitions for device functions like `open` or `llseek`, but this would probably mean the user's custom functions wouldn't interact with WasmFS, which might cause some issues.

Comment (jameshu15869) [reply test/core/test_poll.c:19]: Ah, this makes sense. It looks like the old FS threw `EINVAL` errors when it attempts to call methods that aren't provided. My current approach was to add empty method bodies for `allocFile`, `freeFile`, and `getSize` by default (Since they didn't exist in the old FS) and throw `EINVAL` when the user attempts to call `read` or `write` without passing in the correct values.

However, I'm a little stuck making the WasmFS API `read` and `write` functions match the old API. The original `read` and `write` methods received an entire array object for `buffer` with an offset to start writing at, but the JSIMPL methods take in a pointer to the backend. I tried logging the values and it looks like the old API's `offset` values and the WasmFS API's `buffer` pointer point to different locations. Should they be pointing at the same location in the shared memory between JS and Wasm?

File: test/core/test_poll.c [hunk 7]
```diff
@@ -26,7 +47,12 @@ int main() {
   printf("ret: %d\n", poll(multi, 5, 123));
   printf("errno: %d\n", errno);
   printf("multi[0].revents: %d\n", multi[0].revents == (POLLIN | POLLOUT));
+#if WASMFS
+  // The size of the device is 0 in WasmFS, so the device cannot POLLIN.
+  printf("multi[1].revents: %d\n", multi[1].revents == POLLOUT);
```
Comment (jameshu15869) [inline test/core/test_poll.c:52]: I realized that I forgot to add a comment here, but it looks like `__syscall_poll` checks if a device can POLLIN based on the size of the file. 

Since `getSize()` from JSIMPL wasn't in the original API, my current workaround is to add a default `getSize: () => {}` method when instantiating a device if the user doesn't pass `getSize` to avoid an error for empty devices. Would it make sense for the default method to return a non zero value so `POLLIN` is true (For full compatibility)?

File: test/core/test_poll.c (hunk 6 above)
Comment (kripken) [reply test/core/test_poll.c]: Ah, they might not point to the same place. A JS typed array view has its own internal offset. So if the absolute address of a file in memory was 1000, but we took a view starting from 800, then the offset relative to that view would be 200.

I still hope we can translate the old form to the new? You can do `.byteOffset` on a typed array view to get its offset in the entire space, which might help here.

File: test/unistd/io.c [hunk 8]
```diff
@@ -26,10 +26,8 @@ int main() {
 
     var device = FS.makedev(major++, 0);
     FS.registerDevice(device, {
-      open: function(stream) {
-        stream.payload = [65, 66, 67, 68];
-      },
       read: function(stream, buffer, offset, length, pos) {
+        stream.payload = [65, 66, 67, 68];
```
Comment (jameshu15869) [inline test/unistd/io.c:30]: Is it acceptable to modify existing tests? It looks like the use of `stream.payload` isn't used anywhere else in the code. Attaching a payload to stream in `open` makes compatibility difficult since JS_IMPL doesn't support `open`. This test ensures that WasmFS and the legacy API can use the exact same device operations.

Comment (kripken) [reply test/unistd/io.c:30]: This seems like a reasonable change. But I think we should not reset the payload in each read. Perhaps
```suggestion
        if (!stream.payload) {
          stream.payload = [65, 66, 67, 68];
        }
```

File: src/library_wasmfs.js [hunk 9]
```diff
@@ -392,7 +398,61 @@ FS.createPreloadedFile = FS_createPreloadedFile;
...
+        freeFile: (file) => {
+          wasmFSDeviceStreams[file] = undefined;
+        },
+        getSize: (file) => {},
+        read: (file, buffer, length, offset) => {
+          var bufferArray = Module.HEAP8.subarray(buffer, buffer + length);
+          try {
+            var bytesRead = definedOps.userRead(wasmFSDeviceStreams[file], bufferArray, 0, length, offset);
+          } catch (e) {
+            return -e.errno;
```
Comment (kripken) [inline src/library_wasmfs.js:420]: Interesting, do these functions throw exceptions with an errno? Is there an example of that in the test suite or docs?

**Review by kripken — COMMENTED** [review]
Nice work!

File: src/library_wasmfs.js (hunk 9 above)
Comment (jameshu15869) [reply src/library_wasmfs.js:420]: I did see a test that simulated a "broken device" in `test_unistd_io`, and I think the try/catch is matching the expected behavior here:
https://github.com/emscripten-core/emscripten/blob/8ab7b6d34a53113551fd211c0dd1ea37d2bac2d3/test/unistd/io.c#L53-L61

File: test/unistd/io.c [hunk 10]
```diff
@@ -63,8 +61,11 @@ int main() {
 
     // NB: These are meant to test FS.createDevice specifically,
     //     and as such do not use registerDevice/mkdev
+#if !defined(WASMFS)
+    // WasmFS does not support FS.createDevice()
```
[the remark this answers is not in the document]
Comment (jameshu15869) [reply test/unistd/io.c]: There's no good reason now that I think about it. I think earlier we discussed that since `FS.createDevice` was only used internally and it seemed difficult to match the old API, it might be better to skip. However, some of the other changes I've made since then have made it a little easier to implement `createDevice`, so I've added it to the PR.

File: src/library_wasmfs.js (hunk 9 above)
Comment (kripken) [reply src/library_wasmfs.js:420]: I see, thanks. I guess that API has made sense for JS all along, so good to match it here.

File: test/test_browser.py [hunk 11]
```diff
@@ -2555,6 +2555,8 @@ def test_worker_api_with_pthread_compilation_fails(self):
 
   @also_with_wasmfs
   def test_emscripten_async_wget2(self):
+    if self.get_setting('WASMFS'):
+      self.set_setting('FORCE_FILESYSTEM')
```
Comment (kripken) [inline test/test_browser.py:2559]: Why is this change needed?

**Review by kripken — APPROVED** [review]
Looks good! One last question from me.

Was anything else left to resolve here? (I see the PR is still marked as draft)

File: test/test_browser.py (hunk 11 above)
Comment (jameshu15869) [reply test/test_browser.py:2559]: I think when I initially opened this PR this test was saying `FS is not defined`, though now I suspect that error had more to do with how I was running the tests locally instead of problems with the test itself. It should be fixed now.

**Comment by jameshu15869** [comment]
> Was anything else left to resolve here? (I see the PR is still marked as draft)

There was one comment I had a little earlier (https://github.com/emscripten-core/emscripten/pull/19978#discussion_r1291693460) that had to do with polling a device that I wasn't sure how to resolve, but everything else is resolved

File: test/core/test_poll.c (hunk 7 above)
Comment (kripken) [reply test/core/test_poll.c:40]: Sorry I missed this comment before. Good point about the size...

I'm worried about returning a non-zero value, as that could cause other issues, though my worry is pretty vague.

I lean towards landing this PR as is, maybe just with a TODO on line 51 to mention the issue. @tlively what do you think?
